GH-48172: [Python] Add cp315 to build

paultiq opened the pull request:
### Rationale for this change
Add [Python 3.15](https://docs.python.org/3.15/whatsnew/3.15.html) to the pyarrow nightly builds, so that issues can be detected early in the development cycles of both cpython & pyarrow. 

Note: 
- CFFI and Pandas don't have nightly 3.15 wheels, and are accordingly excluded. 
- Numpy doesn't ship a 3.15 wheel, but is needed to build. It is building from source.

### What changes are included in this PR?
- Adding cp315/cp315t to the build matrix
- Excluding cffi/pandas for 3.15, as the wheels aren't yet available.

### Are these changes tested?
- Yes, tested on a local Linux build using the `ninja-release-python` preset (not maximal)

### Are there any user-facing changes?
No functional changes, other than allowing the user to use Python 3.15. 
* GitHub Issue: #48172

paultiq committed:
ci: add cp315/cp315t to builds.

paultiq committed:
ci: add marker for cffi and pandas for < "3.15". this is temporary until they ship 3.15 wheels.

paultiq commented:
The build was destined for failure, because cp315 is not yet in the manylinux build. I'll reproduce locally and see if there's a graceful way to resolve, along with a clearer failure when PYTHON_ROOT is not found. 

The build failed oddly, but the earlier issue was that the detection of PYTHON_ROOT failed quietly: 
>2025-11-21T08:07:23.9363828Z #20 [stage-0 13/16] RUN PYTHON_ROOT=$(find /opt/python -name cp315-cp315) &&     echo "export PATH=$PYTHON_ROOT/bin:$PATH" >> /etc/profile.d/python.sh
...
>2025-11-21T08:08:01.3779653Z -- Found Python3: /bin/python3.11 (found version "3.11.11") found components: Interpreter

A test should be added to verify PYTHON_ROOT is set / not empty and fail fast:
https://github.com/apache/arrow/blob/2fb2f79a18962875d7f6adc5115666fc0bfea342/ci/docker/python-wheel-manylinux.dockerfile#L110-L111

### Build Failure
Crossbow failed due to 
> lto1: fatal error: bytecode stream in file ‘src/arrow/CMakeFiles/arrow_vendored.dir/Unity/unity_1_c.c.o’ generated with GCC compiler older than 10.0

caused by 
> 2025-11-21T08:08:00.9596950Z -- The C compiler identification is GNU 8.5.0
. 2025-11-21T08:08:01.0402969Z -- The CXX compiler identification is GNU 14.2.1

paultiq committed:
Merge remote-tracking branch 'upstream/main' into cp315_nightly

paultiq commented:
@github-actions crossbow submit wheel-manylinux-2-28-cp315-cp315-amd64

paultiq commented:
Retrying/restarting this, now that 3.15b2 is out and available on ManyLinux.

kou commented:
@github-actions crossbow submit wheel-manylinux-2-28-cp315-cp315-amd64

raulcd commented:
I forgot this issue existed. I took a stab last week as seen here (only local testing):
- https://github.com/apache/arrow/issues/50091#issuecomment-4620455124

I added rustc and cargo to the images but libcst wasn't able to be built for Py3.15 due to PyO3 failures (no 3.15 support). Removing temporarily building libcst and the stubs build I had a green pass for tests.

paultiq commented:
Yup, I see same libcst blocker, cleared by this PR: https://github.com/Instagram/LibCST/pull/1454 (status update/edit: it's now merged, waiting on release, a local test of 1.8.7.dev built successfully)

I got a good build against that PR:
```py
# requirements-wheel-build.txt
libcst @ git+https://github.com/AlexWaygood/LibCST.git@add-315; python_version >= "3.15"
```
```py
# pyproject.toml
    "libcst>=1.8.6; python_version < '3.15'",
    "libcst; python_version >= '3.15'",  # fork doesn't build with the version
```

Happy to close in favor of your PR, but otherwise will just wait for that libcst PR to land, then update this to build against libcst main for 3.15 testing

hugovk commented:
The [Python 3.15 release candidate](https://discuss.python.org/t/python-3-15-0-release-candidate-1-is-here/?u=hugovk) is now out, and https://github.com/Instagram/LibCST/pull/1454 has been merged and released in LibCST v1.9.0: https://github.com/Instagram/LibCST/releases/tag/v1.9.0.

And PyO3 should be ready for 3.15 as well now: https://github.com/PyO3/pyo3/releases/tag/v0.29.0.

paultiq commented:
Could someone run this build? I bumped the libcst dependency, and it built locally. Thx: `@github-actions crossbow submit wheel-manylinux-2-28-cp315-cp315-amd64`

rok commented:
@github-actions crossbow submit wheel-manylinux-2-28-cp315-cp315-amd64

paultiq commented:
The libcst 1.9.0 release worked: build step completes, fails on test. 

A few changes were needed for this next part: 
- PYTHON_IMAGE_TAG: use a "-rc" suffix for 3.15. 
- manylinux 2026.08.08-1: for 3.15rc1.
- disable pandas tests dynamically when pandas is not installed: pandas doesn't yet ship a 3.15 wheel
- housekeeping: reenable cffi since it has a 3.15 wheel, and set a new min numpy wheel for 3.15  

Could someone re-kick that build again? `@github-actions crossbow submit wheel-manylinux-2-28-cp315-cp315-amd64`

rok commented:
@github-actions crossbow submit wheel-manylinux-2-28-cp315-cp315-amd64

rok commented:
@paultiq

paultiq commented:
Great! 

This is ready for a larger crossbow run, to test other platforms + free threading: 
`@github-actions crossbow submit *cp315*`

After that, it's ready for review. This is just to get 3.15 in the nightly builds.

raulcd commented:
@github-actions crossbow submit *cp315*

paultiq commented:
The musllinux are because https://github.com/docker-library/python/pull/1123 dropped ALPINE_LINUX 3.22 builds... so 3.22 is stuck on 3.15.0b2.

I successfully built/tested locally with 3.23. 

## Question 
- Should *this* be a separate PR to update Alpine Linux version, like https://github.com/apache/arrow/pull/47148 (3.18 to 3.22)?
- Should we go to 3.24 (latest) or 3.23 (minimum increase needed)


## Changes needed: 
https://github.com/apache/arrow/blob/e611f48081ec927e18e5bcb72fbd071c15c76b08/.env#L54
https://github.com/apache/arrow/blob/e611f48081ec927e18e5bcb72fbd071c15c76b08/dev/tasks/python-wheels/github.linux.yml#L35
https://github.com/apache/arrow/blob/e611f48081ec927e18e5bcb72fbd071c15c76b08/ci/docker/alpine-linux-3.22-cpp.dockerfile#L19
https://github.com/apache/arrow/blob/e611f48081ec927e18e5bcb72fbd071c15c76b08/ci/docker/alpine-linux-3.22-r.dockerfile#L23

raulcd commented:
On the attempt I did some weeks ago with the change on the following is enough, we don't require changing the default on `.env` + R (we can follow that up on a different issue):
https://github.com/apache/arrow/blob/e611f48081ec927e18e5bcb72fbd071c15c76b08/dev/tasks/python-wheels/github.linux.yml#L35
See:
https://github.com/apache/arrow/pull/50576/changes#diff-58624af974aef2e2980f03f99d2386e78e3fcb1a53d6a4b06612ded49ea8d449

paultiq commented:
Ok, thx, I'll just commit that minimum. What about the .env change? I didn't trace that through.

raulcd commented:
> What about the .env change?

Not necessary, the `.env` is the default and with the change proposed it will already override the default.

paultiq commented:
Thx, and should I keep working on this one / merge in your changes, or do you want to use that other draft?

raulcd commented:
> Thx, and should I keep working on this one / merge in your changes, or do you want to use that other draft?

I am happy to close mine and help you on this one through review, comments if that is ok with you and you want to keep working on it :)
Thanks so much for your work here!

paultiq commented:
Ok, sounds good! I'll review/build the macos and windows versions next, and then request another rebuild.

raulcd commented:
minor note for Windows in order to force the Windows wheels to be rebuilt you'll have to update the following:
https://github.com/apache/arrow/blob/e611f48081ec927e18e5bcb72fbd071c15c76b08/.env#L97-L102

raulcd commented:
@github-actions crossbow submit wheel-musllinux-1-2-cp315-cp315-amd64

paultiq commented:
Could you rerun the full cp315 set? 

- musllinux passed above
- OSX is building locally for cp315 and cp315t: added 3.15 to install_python.sh, and added workaround `--only-binary=:all:` for cp315 (the workaround is to allow builds on cp315 because of one not-yet-available dependency: pyyaml... other platforms don't use --only-binary)
- Windows: I think I have it, but haven't been able to verify locally (local docker issues): https://github.com/apache/arrow/pull/48191/commits/7e5f9008d825f504ce5afed4ddc4b4a4e46cb476

rok commented:
@github-actions crossbow submit *cp315*

paultiq commented:
Ok, almost there... now just macos: problem was that the 3.15.0rc1 led to an attempt to download https://www.python.org/ftp/python/3.15.0rc1/python-3.15.0rc1-macos11.pkg when it should be https://www.python.org/ftp/python/3.15.0/python-3.15.0rc1-macos11.pkg (note the truncated version 3.15.0).

Fixed.

Could someone run: `@github-actions crossbow submit wheel-macos-monterey-cp315-cp315-arm64`

raulcd commented:
@github-actions crossbow submit wheel-macos-monterey-cp315-cp315-arm64

raulcd commented:
@github-actions crossbow submit *cp315*

paultiq commented:
~~Windows builds failed due to github outage~~

raulcd reviewed (approved):
This looks great to me. I think we could merge and have nightly wheels for Python 3.15.
@jorisvandenbossche @AlenkaF @rok  do you want to take a look?

raulcd commented on python/pyproject.toml:
Can you add 3.15 here?

paultiq replied:
Done, added 3.15 to pyproject classifier and 3.15 to the verify release candidate scripts. I believe I got the multiple occurrences.

raulcd reviewed (commented):
After further review, I noticed a couple of minor nits, the classifier and for release preparation a couple places where we should add verification for 3.15 wheels, if you can add those. 

Can you add 3.15 on the Windows verification:
https://github.com/apache/arrow/blob/18146fc78784f482014780417aa5e2546cd76552/dev/release/verify-release-candidate-wheels.bat#L58

And also on the bash verification script (a couple of occurrences):
https://github.com/apache/arrow/blob/18146fc78784f482014780417aa5e2546cd76552/dev/release/verify-release-candidate.sh#L863

Thanks!

rok reviewed (approved):
This looks good to me, but given my lack of knowledge here I defer to @raulcd.
I will only kick off the 3.14 jobs as this PR touches them a little.

rok commented:
@github-actions crossbow submit *cp314*

raulcd commented:
Thanks @rok for kicking off the 3.14 wheels, the failures are related:
```
ERROR: Could not find a version that satisfies the requirement libcst>=1.9.0 (from versions: 0.1, 0.1.1, 0.1.2, 0.1.3, 0.2.0, 0.2.1, 0.2.2, 0.2.3, 0.2.4, 0.2.5, 0.2.6, 0.2.7, 0.3.0, 0.3.1, 0.3.2, 0.3.3, 0.3.4, 0.3.5, 0.3.6, 0.3.7, 0.3.8, 0.3.9, 0.3.10, 0.3.11, 0.3.12, 0.3.13, 0.3.14, 0.3.15, 0.3.16, 0.3.17, 0.3.18, 0.3.19, 0.3.20, 0.3.21, 0.3.22, 0.3.23, 1.8.4, 1.8.5, 1.8.6)
ERROR: No matching distribution found for libcst>=1.9.0
```

paultiq commented:
re: macos errors on amd64: as of libcst 1.9.0, libcst is no longer building a wheel for the only intel builder (macos-13):  https://github.com/Instagram/LibCST/pull/1433

- https://pypi.org/project/libcst/1.8.6/#files has libcst-1.8.6-cp312-cp312-macosx_10_13_x86_64.whl 
- https://pypi.org/project/libcst/1.9.0/#files does not have any macos13 wheels (or any intel macos wheels)

The reason macos worked on cp315 and failed on cp314 is:
- macos is the only platform where --only-binary is passed, which prevents source builds. 
- For cp315, I selectively bypassed this because another dependency (pyyaml, used by libcst) didn't yet have cp315 builds
- Note: On intel, libcst + pyyaml must be built. On arm, only pyyaml is built. 

Options:
- Pin libcst to 1.8.6 for cp314 and below
- Remove libcst dependency
- ?

rok commented:
Pinning libcst is fine imo

paultiq commented:
Will do. I committed, but am running a test (on a crossbow fork) first. I'll ping when it's ready to run. 

This will unlock this PR, but does punt the question of what to do about the `--only-binary` longer term on macos amd64. As it stands, arrow will need to always build libcst from source for the amd64 3.15 builds and above.

paultiq commented:
Could you run wheel-macos-monterey-cp314-cp314-amd64 and wheel-macos-monterey-cp315-cp315-amd64 again?

Thank you

raulcd commented:
@github-actions crossbow submit wheel-macos-monterey-cp314-cp314-amd64 wheel-macos-monterey-cp315-cp315-amd64

raulcd commented:
@github-actions crossbow submit -g wheel

raulcd reviewed (approved):
If no more comments, I'll merge this tomorrow. I will create an issue so we update to RC2 and the final release probably in time for 26.0.0.

jorisvandenbossche commented:
FWIW, pandas should now have nightly wheels for python 3.15, so that might avoid having to skip that

paultiq commented:
Sure, that'd be straightforward, something like this in requirements-test.txt and requirements-wheel-test.txt, and removing the PYARROW_TEST_PANDAS check. 

```
+ --find-links https://pypi.anaconda.org/scientific-python-nightly-wheels/simple/pandas/ # TODO: Remove after cp315 release
+ pandas; python_version < "3.15"
+ pandas>=3.1.0.dev0; python_version >= "3.15" # TODO: Remove after cp315 release
```

This PR or follow-on PR or rc2?

paultiq commented:
@raulcd I can reenable pandas here or in a separate pr as a followup, whichever you prefer. I tested it locally on manylinux and it passed.

raulcd commented:
Can you create an issue to follow it up? I'm happy doing it as a follow up

raulcd commented:
Thanks for working on this @paultiq
